File: solutions/python3/0992.py
class Solution:
    def subarraysWithKDistinct(self, A: List[int], K: int) -> int:
        def subarraysWithAtMostKDistinct(k: int) -> int:
            ans = 0
            count = [0] * (len(A) + 1)
            l = 0
            r = 0

            while r <= len(A):
                if k >= 0:
                    ans += r - l
                    if r == len(A):
                        break
                    count[A[r]] += 1
                    if count[A[r]] == 1:
                        k -= 1
                    r += 1
                else:
                    count[A[l]] -= 1
                    if count[A[l]] == 0:
                        k += 1
                    l += 1

            return ans

        return subarraysWithAtMostKDistinct(K) - subarraysWithAtMostKDistinct(K - 1)

File: solutions/java/0992.java
class Solution {
  public int subarraysWithKDistinct(int[] A, int K) {
    return subarraysWithAtMostKDistinct(A, K) - subarraysWithAtMostKDistinct(A, K - 1);
  }

  private int subarraysWithAtMostKDistinct(int[] A, int K) {
    int ans = 0;
    int[] count = new int[A.length + 1];

    for (int l = 0, r = 0; r <= A.length;) {
      if (K >= 0) {
        ans += r - l;
        if (r == A.length)
          break;
        if (++count[A[r]] == 1)
          --K;
        ++r;
      } else {
        if (--count[A[l]] == 0)
          ++K;
        ++l;
      }
    }

    return ans;
  }
}